### PR TITLE
Ticket 7193

### PR DESCRIPTION
**Pull request by vnepveu**
Add IPv6 support in Sybil attacks protection

### CODE REVIEW

File: src/feature/dirauth/dirvote.c [hunk 1]
```diff
@@ -4198,31 +4199,62 @@ dirserv_get_bandwidth_for_router_kb(const routerinfo_t *ri)
...
+}
+
+/** Helper for sorting: compares two ipv6 routerinfos first by ipv6 address,
+ * and then by descending order of "usefulness"
+ * (see compare_routerinfo_usefulness)
+ **/
+STATIC int
+compare_routerinfo_by_ipv6(const void **a, const void **b)
+{
+  routerinfo_t *first = *(routerinfo_t **)a, *second = *(routerinfo_t **)b;
```
Comment (nmathewson) [inline src/feature/dirauth/dirvote.c:4230]: Let's make these const, just to make sure we're not messing with them.

File: src/feature/dirauth/dirvote.c [hunk 2]
```diff
@@ -4198,31 +4199,62 @@ dirserv_get_bandwidth_for_router_kb(const routerinfo_t *ri)
...
+/** Helper for sorting: compares two ipv4 routerinfos first by ipv4 address,
+ * and then by descending order of "usefulness"
+ * (see compare_routerinfo_usefulness)
  **/
-static int
-compare_routerinfo_by_ip_and_bw_(const void **a, const void **b)
+STATIC int
+compare_routerinfo_by_ipv4(const void **a, const void **b)
 {
   routerinfo_t *first = *(routerinfo_t **)a, *second = *(routerinfo_t **)b;
```
Comment (nmathewson) [inline src/feature/dirauth/dirvote.c:4209]: Let's make these const, just to make sure we're not messing with them.

File: src/feature/dirauth/dirvote.c [hunk 3]
```diff
@@ -4198,31 +4199,62 @@ dirserv_get_bandwidth_for_router_kb(const routerinfo_t *ri)
...
+}
+
+/**
+* Compare routerinfos by descending order of "usefulness" :
+* An authority is more useful than a non-authority; a running router is
+* more useful than a non-running router; and a router with more bandwidth
+* is more useful than one with less.
+**/
+STATIC int
+compare_routerinfo_usefulness(const void **a, const void **b)
```
Comment (nmathewson) [inline src/feature/dirauth/dirvote.c:4248]: Let's have this function take `const routerinfo_t`, since it is never actually passed to smartlist_sort().

**Comment by vnepveu** [comment]
Isn't my assumption that  `global_dirauth_options.AuthDirMaxServersPerAddr == 2 `  a problem ? I tried mocking the corresponding function with little success, so worked with this assumption instead.

**Pull request closed without merging**